ou_dedetai/backup.py

```diff
@@ -0,0 +1,138 @@
+    for src in src_dirs:
+        shutil.copytree(src, Path(dst_dir) / src.name)
+
+
+# FIXME: almost seems like this is long enough to reuse the install_step count in app
+# for a more detailed progress bar
+def backup_and_restore(mode: str, app: App):
+    app.status(f"Starting {mode}…")
+    data_dirs = ['Data', 'Documents', 'Users']
+    backup_dir = Path(app.conf.backup_dir).expanduser().resolve()
+
+    verb = 'Use' if mode == 'backup' else 'Restore backup from'
+    if not app.approve(f"{verb} existing backups folder \"{app.conf.backup_dir}\"?"): #noqa: E501
+        # Reset backup dir.
+        # The app will re-prompt next time the backup_dir is accessed
+        app.conf._raw.backup_dir = None
+
+    # Set source folders.
+    backup_dir = Path(app.conf.backup_dir)
+    try:
+        backup_dir.mkdir(exist_ok=True, parents=True)
+    except PermissionError:
+        verb = 'access'
+        if mode == 'backup':
+            verb = 'create'
+        app.exit(f"Can't {verb} folder: {backup_dir}")
+
+    if mode == 'restore':
+        restore_dir = utils.get_latest_folder(app.conf.backup_dir)
+        restore_dir = Path(restore_dir).expanduser().resolve()
+        # FIXME: Shouldn't this prompt this prompt the list of backups?
+        # Rather than forcing the latest
+        # Offer to restore the most recent backup.
+        if not app.approve(f"Restore most-recent backup?: {restore_dir}", ""):  # noqa: E501
+            # Reset and re-prompt
+            app.conf._raw.backup_dir = None
+            restore_dir = utils.get_latest_folder(app.conf.backup_dir)
+            restore_dir = Path(restore_dir).expanduser().resolve()
+        source_dir_base = restore_dir
+    else:
+        if not app.conf._logos_appdata_dir:
+            app.exit("Cannot backup, Logos installation not found")
+        source_dir_base = Path(app.conf._logos_appdata_dir)
+    src_dirs = [source_dir_base / d for d in data_dirs if Path(source_dir_base / d).is_dir()]  # noqa: E501
+    logging.debug(f"{src_dirs=}")
+    if not src_dirs:
+        app.exit(f"No files to {mode}")
+
+    if mode == 'backup':
+        app.status("Backing up data…")
+    else:
+        app.status("Restoring data…")
+
+    # Get source transfer size.
+    q: queue.Queue[int] = queue.Queue()
+    message = "Calculating backup size…"
+    app.status(message)
+    i = 0
+    t = app.start_thread(utils.get_folder_group_size, src_dirs, q)
+    try:
+        while t.is_alive():
+            i += 1
+            i = i % 20
+            app.status(f"{message}{"." * i}\r")
+            time.sleep(0.5)
+    except KeyboardInterrupt:
+        print()
+        app.exit("Cancelled with Ctrl+C.")
+    t.join()
+    src_size = q.get()
+    if src_size == 0:
+        app.exit(f"Nothing to {mode}!")
+
+    # Set destination folder.
+    if mode == 'restore':
+        if not app.conf.logos_exe:
+            app.exit("Cannot restore, Logos is not installed")
+        dst_dir = Path(app.conf.logos_exe).parent
```

Comment:
Should this use logos app data Dir instead?

Reply:
I had the same question, but I think the Logos *app* should be installed before trying to restore a backup. Otherwise installing Logos over existing data directories might clobber them? Not sure. Also, is `app.conf.logos_exe` generic enough to be defined as the Verbum exe path as well?

Reply:
> Also, is app.conf.logos_exe generic enough to be defined as the Verbum exe path as well?

Yup